File: ONLINE_QUALIFIERS_ROUND_2/6_HOHO.py
from collections import defaultdict, deque


def min_hops(n, routes, source, destination):
    stop_to_routes = defaultdict(set)

    for i, route in enumerate(routes):
        for stop in route:
            stop_to_routes[stop].add(i)

    if source == destination:
        return 0

    graph = defaultdict(set)
    for stop, route_set in stop_to_routes.items():
        route_list = list(route_set)
        for i in range(len(route_list)):
            for j in range(i + 1, len(route_list)):
                a, b = route_list[i], route_list[j]
                graph[a].add(b)
                graph[b].add(a)

    source_routes = stop_to_routes.get(source, set())
    dest_routes = stop_to_routes.get(destination, set())

    if not source_routes or not dest_routes:
        return -1

    visited = set()
    queue = deque()

    for r in source_routes:
        queue.append((r, 1))  
        visited.add(r)

    while queue:
        current_route, hops = queue.popleft()
        if current_route in dest_routes:
            return hops

        for neighbor in graph[current_route]:
            if neighbor not in visited:
                visited.add(neighbor)
                queue.append((neighbor, hops + 1))

    return -1 

n = int(input())
routes = []
for _ in range(n):
    route = list(map(int, input().split()))
    routes.append(route)

source, destination = map(int, input().split())
print(min_hops(n, routes, source, destination))

File: ONLINE_QUALIFIERS_ROUND_2/6_HOHO.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <queue>
#include <set>

using namespace std;
int min_hops(int n, const vector<vector<int>> &routes, int source, int destination)
{
    unordered_map<int, unordered_set<int>> stop_to_routes;
    for (int i = 0; i < n; ++i)
    {
        for (int stop : routes[i])
        {
            stop_to_routes[stop].insert(i);
        }
    }
    if (source == destination)
        return 0;

    unordered_map<int, unordered_set<int>> graph;

    for (const auto &[stop, route_set] : stop_to_routes)
    {
        vector<int> route_list(route_set.begin(), route_set.end());
        for (size_t i = 0; i < route_list.size(); ++i)
        {
            for (size_t j = i + 1; j < route_list.size(); ++j)
            {
                int a = route_list[i], b = route_list[j];
                graph[a].insert(b);
                graph[b].insert(a);
            }
        }
    }

    unordered_set<int> source_routes = stop_to_routes[source];
    unordered_set<int> dest_routes = stop_to_routes[destination];

    if (source_routes.empty() || dest_routes.empty())
        return -1;

    unordered_set<int> visited;
    queue<pair<int, int>> q;

    for (int r : source_routes)
    {
        q.emplace(r, 1);
        visited.insert(r);
    }

    while (!q.empty())
    {
        auto [current_route, hops] = q.front();
        q.pop();

        if (dest_routes.count(current_route))
        {
            return hops;
        }

        for (int neighbor : graph[current_route])
        {
            if (!visited.count(neighbor))
            {
                visited.insert(neighbor);
                q.emplace(neighbor, hops + 1);
            }
        }
    }

    return -1;
}

int main()
{
    int n;
    cin >> n;

    vector<vector<int>> routes(n);
    for (int i = 0; i < n; ++i)
    {
        int stop;
        while (cin.peek() != '\n' && cin >> stop)
        {
            routes[i].push_back(stop);
        }
        cin.ignore();
    }
    int source, destination;
    cin >> source >> destination;

    cout << min_hops(n, routes, source, destination) << endl;
    return 0;
}
